src/isolation.rs

```diff
@@ -0,0 +1,308 @@
+	let dir = Builder::new()
+		.permissions(Permissions::from_mode(0o700))
+		.prefix("uhyve-")
+		.suffix(&Uuid::new_v4().to_string())
+		.tempdir()
+		.ok()
+		.unwrap_or_else(|| panic!("The temporary directory could not be created."));
+
+	let dir_permissions = dir.path().metadata().unwrap().permissions();
+	assert!(!dir_permissions.readonly());
+
+	dir
+}
+
+/// Wrapper around a `HashMap` to map guest paths to arbitrary host paths.
+///
+/// See [crate::hypercall::open] to see this in practice.
+pub struct UhyveFileMap {
+	files: HashMap<String, OsString>,
+}
+
+impl UhyveFileMap {
+	/// Creates a UhyveFileMap.
+	///
+	/// * `mappings` - A list of host->guest path mappings with the format "./host_path.txt:guest.txt"
+	pub fn new(mappings: &[String]) -> UhyveFileMap {
+		if !mappings.is_empty() {
+			UhyveFileMap {
+				files: mappings
+					.iter()
+					.map(String::as_str)
+					.map(Self::split_guest_and_host_path)
+					.map(|(guest_path, host_path)| {
+						(
+							guest_path,
+							fs::canonicalize(&host_path).map_or(host_path, PathBuf::into_os_string),
+						)
+					})
+					.collect(),
+			}
+		} else {
+			UhyveFileMap {
+				files: Default::default(),
+			}
+		}
+	}
+
+	/// Separates a string of the format "./host_dir/host_path.txt:guest_path.txt"
+	/// into a guest_path (String) and host_path (OsString) respectively.
+	///
+	/// * `mapping` - A mapping of the format `./host_path.txt:guest.txt`.
+	fn split_guest_and_host_path(mapping: &str) -> (String, OsString) {
+		let mut mappingiter = mapping.split(":");
+		let host_path = OsString::from(mappingiter.next().unwrap());
+		let guest_path = mappingiter.next().unwrap().to_owned();
+
+		(guest_path, host_path)
+	}
+
+	/// Returns the host_path on the host filesystem given a requested guest_path, if it exists.
+	///
+	/// * `guest_path` - The guest path that is to be looked up in the map.
+	pub fn get_host_path(&mut self, guest_path: &str) -> Option<OsString> {
+		let host_path = self.files.get(guest_path).map(OsString::from);
+		if host_path.is_some() {
+			host_path
+		} else {
+			info!("Guest requested to open a path that was not mapped.");
+			if self.files.is_empty() {
+				info!("UhyveFileMap is empty, returning None...");
+				return None;
+			}
+
+			let requested_guest_pathbuf = PathBuf::from(guest_path);
+			if let Some(parent_of_guest_path) = requested_guest_pathbuf.parent() {
+				info!("The file is in a child directory, searching for a parent directory...");
+				let ancestors = parent_of_guest_path.ancestors();
+				for searched_parent_guest in ancestors {
+					// If one of the guest paths' parent directories (parent_host) is mapped,
+					// use the mapped host path and push the "remainder" (the path's components
+					// that come after the mapped guest path) onto the host path.
+					let parent_host: Option<&OsString> =
+						self.files.get(searched_parent_guest.to_str().unwrap());
+					if let Some(parent_host) = parent_host {
```

Comment:
```suggestion
					if let Some(parent_host) = self.files.get(searched_parent_guest.to_str().unwrap()) {
```

Reply:
Marking as resolved because it is needed for `guest_path_remainder`